src/address_space.rs

```diff
@@ -0,0 +1,307 @@
+//!
+//! To make the abstraction as generic as possible, all the core traits only define methods to
+//! access the address space are defined here, and they never define methods to manage (create,
+//! delete, insert, remove etc) address spaces.  By this way, the address space consumers
+//! (virtio device drivers, vhost drivers and boot loaders etc) may be decoupled from the address
+//! space provider (typically a hypervisor).
+
+use std::cmp::{Eq, Ord, PartialEq, PartialOrd};
+use std::io::{Read, Write};
+use std::ops::{Add, BitAnd, BitOr, Sub};
+use DataInit;
+
+/// Simple helper trait to store a raw address value.
+pub trait AddressValue {
+    /// Type of the address raw value.
+    type V: Copy
+        + PartialEq
+        + Eq
+        + Ord
+        + Add<Output = Self::V>
+        + Sub<Output = Self::V>
+        + BitAnd<Output = Self::V>
+        + BitOr<Output = Self::V>;
+}
+
+/// Trait for address objects, define methods to access and manipulate it.
+///
+/// To simplify the design and implementation, assume the same raw data type could be used to store
+/// address, size and offset for an address space. So the Address trait will be used for address,
+/// size and offset. To ease code review, aliases may be defined though.
+pub trait Address:
+    AddressValue
+    + Sized
+    + Default
+    + Clone
+    + Copy
+    + Eq
+    + PartialEq
+    + Ord
+    + PartialOrd
+    + BitAnd<<Self as AddressValue>::V, Output = Self>
+    + BitOr<<Self as AddressValue>::V, Output = Self>
+{
+    /// Create an address from the raw value.
+    fn new(Self::V) -> Self;
+
+    /// Get the raw value of an address.
+    fn raw_value(&self) -> Self::V;
+
+    /// Returns the bitwise and of the address with the given mask.
+    fn mask(&self, mask: Self::V) -> Self {
+        Self::new(self.raw_value() & mask)
+    }
+
+    /// Returns the offset from this address to the given base address and None if there is
+    /// underflow.
+    fn checked_offset_from(&self, base: Self) -> Option<Self::V>;
+
+    /// Returns the offset from this address to the given base address.
+    /// Only use this when `base` is guaranteed not to overflow.
+    fn unchecked_offset_from(&self, base: Self) -> Self::V {
+        self.raw_value() - base.raw_value()
+    }
+
+    /// Returns the result of the add or None if there is overflow.
+    fn checked_add(&self, other: Self::V) -> Option<Self>;
+
+    /// Returns the result of the base address + the size.
+    /// Only use this when `offset` is guaranteed not to overflow.
+    fn unchecked_add(&self, offset: Self::V) -> Self;
+
+    /// Returns the result of the subtraction or None if there is underflow.
+    fn checked_sub(&self, other: Self::V) -> Option<Self>;
+
+    /// Returns the result of the subtraction.
+    /// Only use this when `other` is guaranteed not to underflow.
+    fn unchecked_sub(&self, other: Self::V) -> Self;
+}
+
+#[macro_export]
+macro_rules! impl_address_ops {
+    ($T:ident, $V:ty) => {
```

Comment:
It will cause compilation errors like:
error: expected expression, found `GuestAddress`
   --> src/address.rs:111:17
    |
111 |                 $T(value)
    |                 ^^ expected expression
    | 
   ::: src/guest_memory.rs:107:1
    |
107 | impl_address_ops!(GuestAddress, u64);
    | ------------------------------------- in this macro invocation